src/PaymentProcessing/SpherePayment.php

```diff
@@ -0,0 +1,257 @@
+                }).catch(function(error) {
+                    alert(error)
+                });
+            }
+            ";
+    }
+
+    /**
+     * Return clinic front specific sphere code in javascript block.
+     *
+     * @return string
+     */
+    private function renderSphereJsClinicFront(): string
+    {
+        return "
+            function sphereSuccess(authCode, transId) {
+                $('#check_number').val(transId);
+                alert(" . xlj('Success') . " + '\\n' + 'Auth: ' + authCode + ' TransId: ' + transId);
+                // core payment process
+                $(\"[name='form_save']\").click();
+            }
+            ";
+    }
+
+    /**
+     * Return core sphere code in javascript block.
+     *   Will employ the front and production status in the rendering.
+     *
+     * @return string
+     */
+    private function renderSphereJsCore(): string
+    {
+        return "
+            function sphereNotSuccess(message) {
+                alert(message);
+            }
+            function randomString(length, chars) {
+                let result = '';
+                for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
+                return result;
+            }
+            function paySphere() {
+                let total = $(\"[name='form_paytotal']\").val();
+                let prepay = $('#form_prepayment').val();
+                if (Number(total) < 1) {
+                    if (Number(prepay) < 1) {
+                        let error = " . xlj('Please enter a payment amount') . "
+                        alert(error);
+                        return false;
+                    }
+                    total = prepay;
+                }
+                $('#form_method').val('credit_card');
+                $('#payTotal').text(total);
+                $('#paymentAmount').val(total);
+                // In case the javascript random generator breaks (ideally use the javascript value in case
+                //  the user makes more than 1 payment from one screen, which would be very unusual to do).
+                let backupTicket = " . js_escape(RandomGenUtils::createUniqueToken(12)) . ";
+                let ticket = randomString(12, '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ');
+                if ((ticket == 0) || (ticket == '') || (ticket.length != 12)) {
+                    error.log('Dynamic javascript ticket creation failed, so using backup ticket.');
+                    ticket = backupTicket;
+                }
+
+                let responseUrl = " . js_escape($this->serverSite) . " + '/sphere/initial_response.php';
+                let cancelUrl = " . js_escape($this->serverSite) . " + '/sphere/initial_response.php?cancel=cancel&front=' + " . js_escape($this->front) . " + '&patient_id_cc=' + " . js_escape($this->patientIdCc) . " + '&csrf_token=' + " . js_escape(CsrfUtils::collectCsrfToken('sphere')) .  ";
```

Comment:
Do we need to add any HMAC signature to make sure the variables were not modified / tampered with here? If another javascript was on this page someone could leave the CSRF alone but modify the other variables.  Maybe I'm being too paranoid here?